### PR TITLE
Checks ISO8601 format for timestamps

### PR DESCRIPTION
**Pull request by dcalvoalonso**
Implements #679.

Unit tests have been added to demonstrate the new functionality for NGSIv1 and NGSIv2.

### CODE REVIEW

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -22,3 +22,4 @@ Fix: updating dependencies due to known vulnerabilities in the previous ones
 Remove: old unused dependencies (sax, grunt, closure-linter-wrapper)
 Fix: mosquitto.conf.example file not found by iot/mosquitto Dockerfile (#554)
 Fix: isDomain is not used anymore for context availability registration (#701)
+Add: checks ISO8601 timeinstants provided by devices (#679)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: Actually #679 is a bug, so "Fix" should be used instead of "Add" to be precise :)

File: lib/errors.js [hunk 2]
```diff
@@ -168,5 +170,10 @@ module.exports = {
         this.name = 'BAD_ANSWER';
         this.message = 'Invalid statusCode in operation [' + operation + ']';
         this.code = 400;
+    },
+    BadTimestamp: function(payload) {
+        this.name = 'BAD_TIMESTAMP';
+        this.message = 'Invalid ISO8601 timestamp[' + payload + ']';
```
Comment (fgalan) [inline lib/errors.js]: Add whitespace just before `[` ?

File: test/unit/ngsiService/active-devices-test.js [hunk 3]
```diff
@@ -352,6 +352,46 @@ describe('Active attributes test', function() {
...
+            modifiedValues = [
+                {
+                    name: 'state',
+                    type: 'Boolean',
+                    value: 'true'
+                },
+                {
+                    name: 'TimeInstant',
+                    type: 'ISO8601',
+                    value: '2015-12-14T08:06:01 00.46Z'
```
Comment (fgalan) [inline test/unit/ngsiService/active-devices-test.js]: What about using exaclty the one that raised the error at #679? I mean:

```
2018-10-05T11:03:56 00:00Z
```

File: test/unit/ngsiv2/ngsiService/active-devices-test.js [hunk 4]
```diff
@@ -217,6 +217,46 @@ describe('Active attributes test', function() {
...
+            modifiedValues = [
+                {
+                    name: 'state',
+                    type: 'Boolean',
+                    value: 'true'
+                },
+                {
+                    name: 'TimeInstant',
+                    type: 'ISO8601',
+                    value: '2015-12-14T08:06:01 00.46Z'
```
Comment (fgalan) [inline test/unit/ngsiv2/ngsiService/active-devices-test.js]: What about using exaclty the one that raised the error at #679? I mean:

```
2018-10-05T11:03:56 00:00Z
```

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (dcalvoalonso) [reply CHANGES_NEXT_RELEASE]: Fixed with ccfb974104b9aceecac889fa524e016c81f8a6e0

File: lib/errors.js (hunk 2 above)
Comment (dcalvoalonso) [reply lib/errors.js]: Fixed with 1cfb789978af5af81078925d636006848a703c19

File: lib/services/ngsi/ngsiService.js [hunk 5]
```diff
@@ -470,9 +470,14 @@ function sendUpdateValueNgsi2(entityName, attributes, typeInformation, token, ca
                         typeInformation,
                         token);
 
-                    if ( (('timestamp' in typeInformation && typeInformation['timestamp'] !== undefined) ? typeInformation.timestamp : config.getConfig().timestamp) &&
-                         ! utils.isTimestampedNgsi2(result)) {
-                        result = addTimestampNgsi2(result, typeInformation.timezone);
+                    if ( ('timestamp' in typeInformation && typeInformation['timestamp'] !== undefined) ? typeInformation.timestamp : config.getConfig().timestamp) {
```
[the remark this answers is not in the document]
Comment (dcalvoalonso) [reply lib/services/ngsi/ngsiService.js]: This works as follows:
1. If `timestamp` field is required by the type (`typeInformation`) or the global configuration (config.getConfig().timestamp), we check if the payload to be sent to the CB is timestamped (`utils.isTimestampedNgsi2`).
2. If it is not timestamped, we add timestamp information (`addTimestampNgsi2`)
3. If it is timestamped, we check that timestamp information is correct (`IsValidTimestampedNgsi2`).

File: lib/services/northBound/restUtils.js [hunk 6]
```diff
@@ -140,6 +141,44 @@ function checkBody(template) {
     };
 }
 
+function IsValidTimestamped(payload) {
```
[the remark this answers is not in the document]
Comment (dcalvoalonso) [reply lib/services/northBound/restUtils.js]: Fixed with 3d126264cb002fbc5a4c4ce2960c126238efc748

File: test/unit/ngsiv2/ngsiService/active-devices-test.js [hunk 7]
```diff
@@ -217,6 +217,46 @@ describe('Active attributes test', function() {
         });
     });
 
+    describe('When the IoTA gets a set of values with a TimeInstant which are not in ISO8601 format', function() {
+        var modifiedValues;
+
+        beforeEach(function(done) {
+            var time = new Date(1438760101468); // 2015-08-05T07:35:01.468+00:00
```
[the remark this answers is not in the document]
Comment (dcalvoalonso) [reply test/unit/ngsiv2/ngsiService/active-devices-test.js]: Fixed with dc0db77cba9d68d8ef89c0bdc22cf8691b7bf112

File: test/unit/ngsiService/active-devices-test.js [hunk 8]
```diff
@@ -352,6 +352,46 @@ describe('Active attributes test', function() {
         });
     });
 
+    describe('When the IoTA gets a set of values with a TimeInstant which are not in ISO8601 format', function() {
+        var modifiedValues;
+
+        beforeEach(function(done) {
+            var time = new Date(1438760101468); // 2015-08-05T07:35:01.468+00:00
```
[the remark this answers is not in the document]
Comment (dcalvoalonso) [reply test/unit/ngsiService/active-devices-test.js]: Fixed with dc0db77cba9d68d8ef89c0bdc22cf8691b7bf112

File: test/unit/ngsiService/active-devices-test.js (hunk 3 above)
Comment (dcalvoalonso) [reply test/unit/ngsiService/active-devices-test.js]: Fixed with 171064daac4c14a3e985453dd040294410ffba2c

File: lib/services/ngsi/ngsiService.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (fgalan) [reply lib/services/ngsi/ngsiService.js]: Thanks for the clarification!

NTC

File: lib/services/northBound/restUtils.js [hunk 9]
```diff
@@ -150,6 +207,12 @@ function isTimestamped(payload) {
     return false;
 }
 
+/**
+ * Checks if timestamp attributes are included in NGSIv1 entities.
+ *
+ * @param {Object} payload         NGSIv1 payload to be analyzed.
+ * @return {Boolean}               true if timestamp attributes are included. false if not.
+ */
 function isTimestampedNgsi2(payload) {
```
Comment (fgalan) [inline lib/services/northBound/restUtils.js]: JavaDoc says "Checks if timestamp attributes are included in NGSIv1 entities" but name of the function mentions NGSIv2...

Is this correct?

**Comment by dcalvoalonso** [comment]
No problem at a!!

**Comment by fgalan** [comment]
LGTM. Thanks!

**Pull request merged**